backend/src/main/java/gov/cdc/usds/simplereport/api/directives/RequiredPermissionsDirectiveWiring.java

```diff
@@ -0,0 +1,238 @@
+            return originalDataFetcher.get(dfe);
+          }
+
+          throw new IllegalGraphqlArgumentException(
+              "Current user does not have permission to supply a non-default value for ["
+                  + environment.getElement().getName()
+                  + "]");
+        };
+
+    overwriteOriginalDataFetcher(environment, newDataFetcher);
+
+    return environment.getElement();
+  }
+
+  @Override
+  public GraphQLFieldDefinition onField(
+      SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment) {
+    GraphQLFieldDefinition fieldDefinition = environment.getElement();
+    var requiredPermissionsBuilder = RequiredPermissions.builder();
+    gatherRequiredPermissions(requiredPermissionsBuilder, environment.getElement());
+    var requiredPermissions = requiredPermissionsBuilder.build();
+
+    var originalDataFetcher = environment.getFieldDataFetcher();
+    DataFetcher<?> newDataFetcher =
+        dfe -> {
+          if (requesterHasRequisitePermissions(dfe, requiredPermissions)) {
+            return originalDataFetcher.get(dfe);
+          }
+
+          var path = dfe.getExecutionStepInfo().getPath();
+          var error =
+              new IllegalGraphqlFieldAccessException(
+                  "Current user does not have permission to request [" + path + "]",
+                  List.of(fieldDefinition.getDefinition().getSourceLocation()),
+                  path.toList());
+
+          // We can either return this error (wrapped in a DataFetcherResult) or throw it. The
+          // former will set the data requested to `null`, and the latter will set the data's
+          // closest non-nullable ancestor to `null`. This directive prefers redacting as little as
+          // possible, so the error is only thrown if the field to which access is being denied is
+          // non-nullable in the schema.
+          if (fieldDefinition.getType() instanceof GraphQLNonNull) {
+            throw error;
+          }
+
+          return DataFetcherResult.newResult().error(error).build();
+        };
+
+    overwriteOriginalDataFetcher(environment, newDataFetcher);
+
+    return environment.getElement();
+  }
+
+  private boolean argumentHasDefaultOrNullValue(
+      DataFetchingEnvironment dfe, GraphQLArgument argument) {
+    var argValue = dfe.getArgument(argument.getName());
+
+    return argValue == null || Objects.equals(argValue, argument.getDefaultValue());
+  }
+
+  private boolean requesterHasRequisitePermissions(
+      DataFetchingEnvironment dfe, RequiredPermissions requiredPermissions) {
+    return getSubjectFrom(dfe)
+        .map(
+            subject ->
+                satisfiesRequiredPermissions(
+                    requiredPermissions, subject, dfe.getExecutionStepInfo().getPath()))
+        .orElse(false);
+  }
+
+  private Optional<Subject> getSubjectFrom(DataFetchingEnvironment dfe) {
+    return Optional.ofNullable(dfe.getContext())
```

Comment:
I see why you would do this, though part of me just wants to say

    GraphQLContext c = dfe.getContext();
    return c.getSubject();

😁